July-Morning/assignment2/Assignment_2.cpp

```diff
@@ -0,0 +1,166 @@
+#include <iostream>
+#include <string>
+#include <list>
+
+template <typename T> struct node
+{
+    T key;
+    node * left;
+    node * right;
+    node (T init_key)
+    {
+        key = init_key;
+        left = NULL;
+        right = NULL;
+    }
+};
+
+// Here we assume that all keys in the tree are unique
+template <typename T> class binary_tree 
+{
+    public:
+    
+        node<T> * root;
+    
+        binary_tree() : root(NULL) {}
+        
+        ~binary_tree()
+        {
+            if (root != NULL)
+                destroy(root);
+        }
+        
+        void destroy (node<T> * s_root) 
+        {
+            if (s_root->left != NULL)
+                destroy (s_root->left);
+            if (s_root->right != NULL)
+                destroy (s_root->right);
+            delete s_root;   
+        }
+        
+        bool is_child (T& key, node<T> * parent) 
+        {
+            if (parent == NULL)
+                return false;
+            if (parent->left != NULL)    
+                if (parent->left->key == key)
+                    return true;
+            if (parent->right != NULL)    
+                if (parent->right->key == key)
+                    return true;        
+            return (is_child(key, parent->left) || is_child(key, parent->right));
+        }
+        
+        // Find ancestors of the node with a given key
+        // Return value:
+        // true - if ancestors were found successfully
+        // false - if tree is empty or node with given key does not exist
+        bool find_ancestors (node<T> * s_root, T& key, std::list<node<T> *>& ancestors) 
+        {
+            if (s_root == NULL)
+                return false;
+            if (s_root->key == key)
+                return true;
+            bool is_child = (find_ancestors(s_root->left, key, ancestors) || find_ancestors(s_root->right, key, ancestors));    
```

Comment:
No need for parenthesis after `=`:
```C++
bool is_child = find_ancestors(s_root->left, key, ancestors) || find_ancestors(s_root->right, key, ancestors); 
```
Variable name `is_child` helps understand the meaning of the result of `find_ancestors(left) || find_ancestors(right)`. :+1: 

Only note that it the same name as the `is_child` function. It is ok in this code because of hiding (the local variable `is_child` hides the function), but it still has downsides, for example:
- It can be confusing for the reader which may confuse and think that this is the result of the same function call
- It can result in funny bugs, for example: you delete the definition `bool is_child = ...`. Then, the line `if (is_child)` works and evaluated as true because `is_child` refers to the function pointer... better to avoid in advance :)